File: sdk/go/protomcp/local_middleware.go
package protomcp

import (
	"sort"
	"sync"
)

// LocalMiddlewareDef holds a registered local (in-process) middleware definition.
type LocalMiddlewareDef struct {
	Priority int
	Handler  func(ctx ToolContext, toolName string, args map[string]interface{}, next func(ToolContext, map[string]interface{}) ToolResult) ToolResult
}

var (
	localMwRegistry []LocalMiddlewareDef
	localMwMu       sync.Mutex
)

// LocalMiddleware registers a local middleware that wraps tool handlers.
// Priority determines execution order: lower = outermost.
func LocalMiddleware(priority int, handler func(ctx ToolContext, toolName string, args map[string]interface{}, next func(ToolContext, map[string]interface{}) ToolResult) ToolResult) {
	localMwMu.Lock()
	defer localMwMu.Unlock()
	localMwRegistry = append(localMwRegistry, LocalMiddlewareDef{
		Priority: priority,
		Handler:  handler,
	})
}

// GetLocalMiddleware returns middleware sorted by priority (lowest first = outermost).
func GetLocalMiddleware() []LocalMiddlewareDef {
	localMwMu.Lock()
	defer localMwMu.Unlock()
	result := make([]LocalMiddlewareDef, len(localMwRegistry))
	copy(result, localMwRegistry)
	sort.Slice(result, func(i, j int) bool {
		return result[i].Priority < result[j].Priority
	})
	return result
}

// ClearLocalMiddleware removes all registered local middleware.
func ClearLocalMiddleware() {
	localMwMu.Lock()
	defer localMwMu.Unlock()
	localMwRegistry = nil
}

// BuildMiddlewareChain wraps the handler with all registered local middleware.
// Returns a function with the same signature as the original handler.
func BuildMiddlewareChain(toolName string, handler func(ToolContext, map[string]interface{}) ToolResult) func(ToolContext, map[string]interface{}) ToolResult {
	middlewares := GetLocalMiddleware()
	if len(middlewares) == 0 {
		return handler
	}

	chain := handler
	// Build from innermost (highest priority) to outermost (lowest priority).
	// GetLocalMiddleware returns sorted lowest-first, so we reverse iterate.
	for i := len(middlewares) - 1; i >= 0; i-- {
		mwHandler := middlewares[i].Handler
		nextFn := chain
		chain = func(ctx ToolContext, args map[string]interface{}) ToolResult {
			return mwHandler(ctx, toolName, args, func(c ToolContext, a map[string]interface{}) ToolResult {
				return nextFn(c, a)
			})
		}
	}
	return chain
}

File: sdk/python/src/protomcp/local_middleware.py
from dataclasses import dataclass
from typing import Any, Callable

_local_mw_registry: list["LocalMiddlewareDef"] = []

@dataclass
class LocalMiddlewareDef:
    priority: int
    handler: Callable  # (ctx, tool_name, args, next_handler) -> ToolResult

def local_middleware(priority: int = 100):
    """Register a local (in-process) middleware that wraps tool handlers."""
    def decorator(func: Callable) -> Callable:
        _local_mw_registry.append(LocalMiddlewareDef(priority=priority, handler=func))
        return func
    return decorator

def get_local_middleware() -> list[LocalMiddlewareDef]:
    """Return middleware sorted by priority (lowest first = outermost)."""
    return sorted(_local_mw_registry, key=lambda m: m.priority)

def clear_local_middleware():
    _local_mw_registry.clear()

def build_middleware_chain(tool_name: str, handler: Callable) -> Callable:
    """Build a chain wrapping handler. Returns callable: (ctx, args_dict) -> result.

    The handler is called as handler(ctx=ctx, **args) if ctx is not None, else handler(**args).
    Each middleware receives (ctx, tool_name, args, next_handler) where next_handler is (ctx, args) -> result.
    """
    middlewares = get_local_middleware()

    def final_handler(ctx, args: dict):
        if ctx is not None:
            return handler(ctx=ctx, **args)
        return handler(**args)

    chain = final_handler
    for mw in reversed(middlewares):
        mw_handler = mw.handler
        def make_link(mw_fn, next_fn):
            def link(ctx, args: dict):
                def call_next(c, a):
                    return next_fn(c, a)
                return mw_fn(ctx, tool_name, args, call_next)
            return link
        chain = make_link(mw_handler, chain)

    return chain
